### ISSUE TITLE
not storing logs at leader

### ISSUE THREAD

**Issue by magnoyu** [hashicorp/raft#507]
Hi, I have a quick question.. let's say I have a 5 nodes clusters. Let's say I don't store logs at the leader, but I store them at all the 4 followers. What is the implication during recovery? As an example, let's say we have persisted log

Leader Node 1 log: empty
Node 2 log: index 21
Node 3 log: index 32
Node 4 log: index 45
Node 5 log: index 55

Let's say the leader has received up to index 55 and never store any logs. 
In terms of commit, the cluster has committed and processed up to index 45. 
Let's say the cluster are all terminated and now we want to recover.
Is there a way to recover from index 45 although that's only the 2nd (out of 5) latest persisted log, instead of from index 32, which I think normally that's what the cluster would have done? 
The hypothesis is, follower nodes cannot have received any log without the leader first receiving it. So I could have copied the node 5 logs to node 1 and start the recovery... but I wonder if this is something configurable such that I don't need to do the manual copy.

**Comment by thefallentree** [comment]
@banks  could you help out?

**Comment by banks** [comment]
Hey @magnoyum as far as I can tell the state you describe is not possible for our Raft implementation and possibly not by any correct raft implementation.

In normal operation your example couldn't happen because Raft explicitly requires that a new leader is the most up-to-date one in the cluster so Node 1 would never have won an election and become leader in your example since the other peers would never have voted for a server with a shorter log than them.

But you asked about what if all servers stopped and now need to come back up. In this case Raft's guarantees are broken and any recovery process must by definition have at least some risk of data loss. That said, our raft implementation couldn't get in this state still as far as I can tell because we write logs to the leaders disk _and wait for fsync_ before we replicate to any peers so if all 5 servers stop at the same moment, the node that was the leader at the point they stopped must have the most up to date log and can't be missing any log entries that do exist on other peers.

Now if the leaders disk is not _really_ doing an fsync when the OS requests it then that's a different story, but generally all durability/linearizability guarantees are broken!

But let's say that _somehow_ you got in the state you described with all servers down and the old leader "behind" on disk. Then what would happen when you start each process up is that the old leader would never be re-elected due to the voting protocol and one of the other replicas would become the new leader and so the committed state would not be lost.

Was this a theoretical question or are you trying to understand a real issue you experienced with this library?

**Comment by banks** [comment]
Ah I see.

Yeah every commit goes to leader logs first before being replicated in our implementation. It's technically possible in Raft to write the leader's log in parallel with replication - this is an optimization detailed in Diego's full Raft thesis. but to implement it would mean major changes to this library. It's something that while I'd personally be interested in doing it, I don't think it's enough of a pain point for our internal use-cases to warrant the large engineering effort it will take to test and be confident in a major change to our raft lib!

Etcd's raft library IIRC can make use of that optization so that might be a consideration.

All that said, if you don't fsync every log write then the correctness of Raft is much harder to reason about. All of the formal proofs of raft's correctness and all of the recovery procedures are based on the assumption that disk writes once confirmed are never lost.

Depending on how important your data is, you _might_ be OK with occasional data loss.

In other words, there is no correct (in the sense of linearizable under all situations) way to use this library without Fsync on every single LogStore append. If you choose to anyway for performance reasons that may be suitable but it goes against all the design assumptions in the protocol and library so you will have to do a lot of reasoning for yourself to understand the risks and failure modes involved!

**Comment by thefallentree** [comment]
@banks  thanks for the detailed explanation.  We were seeing some notable performance improvement in the changes we made , but unsure about the risk induced. (To clarify: we changed so Leader doesn't do fsync , only followers do)  and it brings about 10%-15% latency gain

My question is, what is the best way for us to evaluate the risk introduced? is there some automated test/process we could run to check if we will see any data loss with this change?

**Comment by banks** [comment]
> what is the best way for us to evaluate the risk introduced?

That's quite hard to answer! As far as I know Raft has only been proven correct assuming every log append is actually persisted on disk. It is only specified for a model where a node that has crashed and restarted comes back with either all the state it has previously acknowledged or none at all (and so is the same as adding a brand new node to the cluster).

To evaluate the possible outcomes of changing one of those fundamental assumptions requires very careful analysis and testing. Depending on the level of confidence you need, you might want to re-model in TLA+ to explore the full space of outcomes that might arise. You may wish to do Jepsen style "chaos" testing where you run simulated workloads with lots of node failures that loose disk etc.

Or you can choose to say "seems fine, the chance of a crash that looses unsynced data is low enough we'll just wait and see and recover from a backup if we really need to". It really depends on the criticality of your system 😄!

FWIW not fsyncing logs is less obviously bad than non fsyncing the Stable store where term/vote information is stored. If that is style doing fsync then at least you don't have trivial split brain issues where multiple nodes can be elected leader in the same term etc.

But it is _more_ risky with our Raft implementation than with other ones because we _assume_ that once leader has dispatched logs that they are "safe" and automatically count the leader as one of the quorum when we decide if a log is committed. If it's actually not been synched and is lost then it's possible that that write will be lost forever even though we might have "acked" back to the client that wrote violating Linearizability. If that's not a big deal in your case then fine but it certainly would fail a Jepsen style test for linerizability guarantees.

To actually observe a violation in real life might actually be quite hard but it will be possible. Off the top of my head something like this _could_ occur:

Assume there is a 5 node cluster of servers A, B, C, D, E. Quorum size is 3, current leader is A.

1. Nodes D and E get partitioned away The last index in their logs is 10. They can't do anything as they are in a minority partition.
2. Nodes A, B, C continue to commit entries 11, 12, 13, but the leader A is not syncing them to disk.
3. The datacenter power goes out and all servers stop
4. Servers come back on line slowly as different racks are powered up
5. Servers A, D and E come back - the partition is healed as the network switch has reset
6. Server A recovers only the logs that we fsynced which was only up to entry 9
7. D and E have logs up to 10
8. The three servers are enough to form a Quorum so they do and elect D as the new leader since it has the most recent logs.
9. Server A the old leader is now a follower and replicates 10 again from the new leader
10. The cluster is up and running just fine with no errors in logs... but it lost logs 11, 12, 13!
11. Eventually B and C come back but their 11, 12, 13 entries are for an older term than the new leader and so they _will be discarded_ because raft assumes that they can't possibly ever have been committed without being persisted on a quorum of disks...

That's one example off the top of my head. It might be a risk you're willing to take but it shows why this isn't "correct" in all possible cases! Remember too that "network partitions" can really mean any sort of transient network failure so even if this power sequence seems far fetched, it's hard to reason about every possible type of failure that could lead to an equivalent outcome - that's what model checking these algorithms and formal proofs are good for!

Does this help?

**Comment by banks** [comment]
FWIW I think a raft implementation that writes the leader logs in parallel would perform _closer_ to your no-fsync benchmark because, assuming that disk are only sometimes slow to fsync, on those occasions the latency is bound by the slowest in _any_ quorum - so if the leader is slower than another follower you don't have to wait for the leader.

That's not to mention that you'll automatically get a major latency improvement from that optimisation anyway because it's onw only 1 RTT and 1 fsync delay per commit instead of 1 RTT and 2 fsync delays.

But, as mentioned, that would come close to a full rewrite of the library so is unlikely to be something that is prioritized soon! If you are interested, you may want to consider alternative implementations!

If you do decide to go ahead with no fsync, you might consider just dropping fsync on all logs not just the leader for a much bigger improvement. You've already given up on Linearizability/strict consistency anyway so it's not that likely to be much worse if none of the servers fsync logs and you'd get a much bigger performance increase!

**Comment by banks** [comment]
I'm going to close this as I think the original question was answered. The (correct) optimization discussed here about committing leaders logs in parallel is a topic I'm still interested in and have had some new ideas about but I will continue with those in a new issue as it's not the same topic that was raise here.
